src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -38,7 +38,7 @@ public override bool CanConvert(Type typeToConvert)
         /// <param name="typeToConvert">The <see cref="Type"/> being converted.</param>
         /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
         /// <returns>The value that was converted.</returns>
-        public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options);
+        public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions? options);
```

Comment:
This should not be nullable.
```suggestion
        public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options);
```

Reply:
Thinking about this some more. The returned type can be null. Similarly, the T on the Write can be null as well.

So, this is how I would annotate it:
```C#
[return: MaybeNull]
public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options);
public abstract void Write(Utf8JsonWriter writer, [AllowNull] T value, JsonSerializerOptions options);
```

Reply:
Shouldn't the return value be dictated by the nullability of `T`? So if I create a JsonConverter<string> I know I can't get null strings? Same thing for `T value`... if I want to write nullable T's I should instantiate the `JsonConverter<T>` with a nullable `T`.

Reply:
> Shouldn't the return value be dictated by the nullability of `T`?

What is the nullability of `T` if:
- `T` is a reference type (like string, class Foo)?
- `T` is a value type (like int32, struct Foo, DateTime)?
- `T` is a nullable value type (like int32?, struct Foo?, DateTime?)?

We have an internal converter that inherits from `JsonConverter<string?>` which can return a null string on `Read`. Should we not annotate the public base type to correctly show that `Read` can return null for that?

Reply:
> We have an internal converter that inherits from JsonConverter<string?> which can return a null string on Read.

Yes but I don’t think the abstract class should have MaybeNull nor AllowNull. I think whatever the nullability of T used to create the JsonConverter will dictate whether it allows nulls or not.

So if you have `JsonConverter<string>` the nullability of `T` is not nullable. 

So that’s what I mean, if people expect to have nulls they should create the converter with a nullable reference type. For example `JsonConverter<string?>`

The only confusing piece here is that people don’t control the payload, so they could be getting a null on the payload even if the converter was created with a not nullable `T`, however I don’t think we should make people pay the cost of checking for null if they use a non nullable reference type as `T`.

@stephentoub, thoughts?

Reply:
> Yes but I don’t think the abstract class should have MaybeNull nor AllowNull. I think whatever the nullability of T used to create the JsonConverter will dictate whether it allows nulls or not.

I haven't fully grasped nullability (especially for generics), so I am fine with deferring to you on this.

That said, let me see if I can understands things a bit better.

Here's a sample where the Read method is returning a null reference type. Should/will the compiler warn the author of the converter that they are returning a null? And then, to fix that they would have to change `JsonConverter<Foo>` to `JsonConverter<Foo?`>?

```C#
public class Foo
{
}

public class MyBadConverter : JsonConverter<Foo>
{
    public override Foo Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return null;
    }

    // override Write method
}
```

Reply:
> Should/will the compiler warn the author of the converter that they are returning a null?

Yes, it will.  The method is typed to return the non-nullable `Foo`.  The compiler will then yell at you for returning something that may be null (it doesn't distinguish definitely null from maybe null, so `return null` is maybe null).

> then, to fix that they would have to change JsonConverter<Foo> to JsonConverter<Foo?>

Correct.  Then the return type of `Read` would need be `Foo?`, and returning `null` is fine, because `null` is a valid value for a nullable type.

Reply:
Sound like it is not [return: MaybeNull] or [AllowNull] situation

Reply:
> Sounds like it is not [return: MaybeNull] or [AllowNull] situation

Yep.

**Question:** When would we use `[return: MaybeNull]`?

The instances I found were in Find/FindLast on types like List/Linq where we explicitly return `default!` in some cases. Is that when we would do so?
https://github.com/dotnet/runtime/blob/b9b60ae9ebcf5f3674e0f72f79ab6f82b5c7c080/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L417-L418
https://github.com/dotnet/runtime/blob/b9b60ae9ebcf5f3674e0f72f79ab6f82b5c7c080/src/libraries/System.Linq/src/System/Linq/Single.cs#L126-L127

Reply:
Exactly, if method might return default value for non nullable ref types